File: sync-domain/src/main/java/dev/bee/kanjianki/syncdomain/ProviderCardPolicy.java
package dev.bee.kanjianki.syncdomain;

import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class ProviderCardPolicy {
    private static final String STABILITY = "stability";
    private static final String DIFFICULTY = "difficulty";
    private static final String RETRIEVABILITY = "retrievability";
    private static final Pattern FSRS_DATA_VALUE = Pattern.compile(
            "(?:\"|')?(stability|difficulty|retrievability|s|d|r)(?:\"|')?\\s*[:=]\\s*\"?([-+]?[0-9]+(?:\\.[0-9]+)?)\"?",
            Pattern.CASE_INSENSITIVE
    );
    private static final Pattern FINITE_DOUBLE_VALUE = Pattern.compile("[-+]?(?:[0-9]+(?:\\.[0-9]*)?|\\.[0-9]+)(?:[eE][-+]?[0-9]+)?");

    private ProviderCardPolicy() {
    }

    public static boolean shouldReportCardProgress(int scanned, int total) {
        if (scanned <= 0 || scanned == total || total <= 100) {
            return true;
        }
        if (scanned <= 10) {
            return true;
        }
        return scanned % (total <= 1000 ? 10 : 50) == 0;
    }

    public static FsrsMemoryState fsrsMemoryState(
            String fsrsStability,
            String legacyStability,
            String fsrsDifficulty,
            String legacyDifficulty,
            String fsrsRetrievability,
            String legacyRetrievability,
            String serializedData
    ) {
        Double stability = firstFiniteDouble(fsrsStability, legacyStability);
        Double difficulty = firstFiniteDouble(fsrsDifficulty, legacyDifficulty);
        Double retrievability = firstFiniteDouble(fsrsRetrievability, legacyRetrievability);
        if (stability != null || difficulty != null || retrievability != null) {
            return new FsrsMemoryState(stability, difficulty, retrievability);
        }
        return parseFsrsData(serializedData);
    }

    static FsrsMemoryState parseFsrsData(String data) {
        if (data == null || data.trim().isEmpty()) {
            return FsrsMemoryState.EMPTY;
        }
        Double stability = null;
        Double difficulty = null;
        Double retrievability = null;
        Matcher matcher = FSRS_DATA_VALUE.matcher(data);
        while (matcher.find()) {
            Double value = parseDouble(matcher.group(2));
            if (value == null) {
                continue;
            }
            String key = matcher.group(1).toLowerCase(Locale.ROOT);
            if (STABILITY.equals(key) || "s".equals(key)) {
                stability = value;
            } else if (DIFFICULTY.equals(key) || "d".equals(key)) {
                difficulty = value;
            } else {
                retrievability = value;
            }
        }
        return new FsrsMemoryState(stability, difficulty, retrievability);
    }

    static Double parseDouble(String value) {
        if (value == null || !FINITE_DOUBLE_VALUE.matcher(value).matches()) {
            return null;
        }
        double parsed = Double.parseDouble(value);
        return Double.isInfinite(parsed) ? null : parsed;
    }

    private static Double firstFiniteDouble(String firstValue, String fallbackValue) {
        Double value = parseDouble(trim(firstValue));
        return value == null ? parseDouble(trim(fallbackValue)) : value;
    }

    private static String trim(String value) {
        return value == null ? null : value.trim();
    }

    public record FsrsMemoryState(Double stability, Double difficulty, Double retrievability) {
        private static final FsrsMemoryState EMPTY = new FsrsMemoryState(null, null, null);
    }
}

File: sync-domain/src/main/kotlin/dev/bee/kanjianki/syncdomain/ProviderCardPolicy.kt
package dev.bee.kanjianki.syncdomain

import java.util.Locale
import java.util.regex.Pattern

class ProviderCardPolicy private constructor() {
    @JvmRecord
    data class FsrsMemoryState(
        val stability: Double?,
        val difficulty: Double?,
        val retrievability: Double?,
    )

    companion object {
        private const val STABILITY = "stability"
        private const val DIFFICULTY = "difficulty"
        private const val RETRIEVABILITY = "retrievability"
        private val EMPTY = FsrsMemoryState(null, null, null)
        private val FSRS_DATA_VALUE = Pattern.compile(
            "(?i)['\"]?(stability|difficulty|retrievability|s|d|r)['\"]?\\s*[:=]\\s*['\"]?([^\\s,'\"{}\\[\\]]+)"
        )

        /**
         * True when the card is suspended.
         *
         * The rule is *any* negative queue, not Anki's narrower `queue == -1`. A
         * user-buried or sibling-buried card is not active study material either,
         * and treating only `-1` as suspended would let a buried card count as
         * mature support for a kanji the learner is not actually being shown.
         */
        @JvmStatic
        fun isSuspendedQueue(rawQueue: Long): Boolean = rawQueue < 0L

        /**
         * A configured-model card is kept only at template ordinal 0, Kani's front
         * template. Accepting a reverse or extra template would let one note supply
         * two independent-looking passes for the same kanji.
         */
        @JvmStatic
        fun isAcceptedTemplateOrd(rawTemplateOrd: Long): Boolean = rawTemplateOrd == 0L

        /**
         * Converts Anki's raw `ivl` into the whole days the snapshot carries.
         *
         * Anki stores a **negative** `ivl` to mean seconds, for sub-day learning
         * cards. Passed through unchanged that becomes a negative day count, and
         * maturity arithmetic compares day counts against a positive threshold — so
         * a card answered minutes ago would sort as further from mature than one at
         * a genuine long interval, and any consumer that sums intervals would be
         * pulled downward by it. Sub-day therefore floors to 0.
         */
        @JvmStatic
        fun intervalDays(rawInterval: Long): Int {
            if (rawInterval <= 0L) return 0
            return rawInterval.coerceAtMost(Int.MAX_VALUE.toLong()).toInt()
        }

        /**
         * Clamps a provider counter (`reps`, `lapses`) into the snapshot's `Int`
         * field. Negative counters are not meaningful, so they floor at 0; the top
         * saturates rather than wrapping, because a wrapped count would read as a
         * *negative* number of lapses.
         */
        @JvmStatic
        fun counter(rawValue: Long): Int =
            rawValue.coerceIn(0L, Int.MAX_VALUE.toLong()).toInt()

        /**
         * Clamps a signed provider value (`queue`, `due`) into the snapshot's `Int`
         * field. Unlike [counter] the sign is preserved: `queue` uses negatives for
         * suspended and buried, and `due` is relative in some queues.
         */
        @JvmStatic
        fun signed(rawValue: Long): Int =
            rawValue.coerceIn(Int.MIN_VALUE.toLong(), Int.MAX_VALUE.toLong()).toInt()

        @JvmStatic
        fun shouldReportCardProgress(scanned: Int, total: Int): Boolean {
            if (scanned <= 0 || scanned == total || total <= 100) {
                return true
            }
            if (scanned <= 10) {
                return true
            }
            return scanned % (if (total <= 1000) 10 else 50) == 0
        }

        @JvmStatic
        fun fsrsMemoryState(
            fsrsStability: String?,
            legacyStability: String?,
            fsrsDifficulty: String?,
            legacyDifficulty: String?,
            fsrsRetrievability: String?,
            legacyRetrievability: String?,
            serializedData: String?,
        ): FsrsMemoryState {
            val stability = firstFiniteDouble(fsrsStability, legacyStability)
            val difficulty = firstFiniteDouble(fsrsDifficulty, legacyDifficulty)
            val retrievability = firstFiniteDouble(fsrsRetrievability, legacyRetrievability)
            if (stability != null || difficulty != null || retrievability != null) {
                return FsrsMemoryState(stability, difficulty, retrievability)
            }
            return parseFsrsData(serializedData)
        }

        @JvmStatic
        fun parseFsrsData(data: String?): FsrsMemoryState {
            if (data == null || data.trim().isEmpty()) {
                return EMPTY
            }
            var stability: Double? = null
            var difficulty: Double? = null
            var retrievability: Double? = null
            val matcher = FSRS_DATA_VALUE.matcher(data)
            while (matcher.find()) {
                val value = parseDouble(matcher.group(2)) ?: continue
                when (matcher.group(1).lowercase(Locale.ROOT)) {
                    STABILITY, "s" -> stability = value
                    DIFFICULTY, "d" -> difficulty = value
                    else -> retrievability = value
                }
            }
            return FsrsMemoryState(stability, difficulty, retrievability)
        }

        @JvmStatic
        fun parseDouble(value: String?): Double? {
            return value?.trim()?.toDoubleOrNull()?.takeIf { it.isFinite() }
        }

        private fun firstFiniteDouble(firstValue: String?, fallbackValue: String?): Double? {
            val value = parseDouble(trim(firstValue))
            return value ?: parseDouble(trim(fallbackValue))
        }

        private fun trim(value: String?): String? = value?.trim()
    }
}
